### PR TITLE
Make analysis reduction a Mantid algorithm

### LINKED ISSUE #127: Convert analysis_reduction.py script into a Mantid Algorithm
**Work Description**
Currently the analysis reduction routine is done by running the analysis_reductin.py script. This script calls other scripts for validation of inputs and preprocessing workspaces.
The entire routine would be better organized and cleaner if it was turned into a mantid algorithm. 
This work will make it easier to test, maintain and use the bootstrap routines in the future.

### PR DESCRIPTION
**Pull request by GuiMacielPereira**
**Description of work:**
Currently the analysis scripts are based on a terrible design for running the algorithms.
The inputs are passed as a single object containing all the input arguments, which then are used by functions all the way down the reduction until the results are created as an object at the end of the routine.

Input validation is poor, the procedures defined are highly confusing to understand and the main analysis  routine contains functions with many shared arguments that should be in principle contained within the same object.

**Proposed Solution**
Both the analysis and yfit routine should be made into a Mantid Algorithm. 
The design means that certain arguments are easily available to function throughout the algorithm and it's simpler to keep track of the state as the algorithm progresses.
Dealing with inputs, input validation, constructing procedures and extracting outputs will be highly simplified and made robust.

In particular, VESUVIO has many diferent ways of chaining routines such that the fit parameters of profiles from backscattering detectors should be automatically be the initial guess for the fitting parameters of the forward scattering detectors. An OOP design will make this transition much smoother and whole lot more safe than it is currently.

Steps until Mantid algorithm:

- [x] Transition `analysis_reduction.py` to an object (non-Mantid yet)
- [x] Run system tests with new format and check that they pass
- [ ] Create new input/output interface (replace old way of using a script for inputs)
- [x] Re-write `analysis_routines.py` to use new design
- [x] Add chain between routines functunality
- [ ] Make object a Mantid Algorithm
- [ ] Do input and output validation

**To test:**

Fixes #127 .

### CODE REVIEW

File: src/mvesuvio/oop/NeutronComptonProfile.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+
+
+
+class NeutronComptonProfile:
```
Comment (ajjackson) [inline src/mvesuvio/oop/NeutronComptonProfile.py:4]: How about implementing this as

```python
from dataclasses import dataclass

@dataclass(frozen=True)
class NeutronComptonProfile:
    mass: float
    label: str
    intensity: float
    width: float
    center: float
    intensity_bounds: tuple[float, float]
    width_bounds: tuple[float, float]
    center_bounds: tuple[float, float]

```

(The type annotations are a blind guess as I'm not familiar with the rest of the code...)

File: src/mvesuvio/oop/analysis_helpers.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+
+def gaussian(x, sigma):
+    """Gaussian function centered at zero"""
+    gaussian = np.exp(-(x**2) / 2 / sigma**2)
+    gaussian /= np.sqrt(2.0 * np.pi) * sigma
+    return gaussian
+
+
+def lorentizian(x, gamma):
```
Comment (ajjackson) [inline src/mvesuvio/oop/analysis_helpers.py:152]: Typo here, should be `lorentzian`

File: src/mvesuvio/oop/AnalysisRoutine.py [hunk 3]
```diff
@@ -0,0 +1,1066 @@
...
+                print("Skipped spectra.")
+            else:
+                with np.printoptions(
+                    suppress=True, precision=4, linewidth=200, threshold=sys.maxsize
+                ):
+                    print(specFitPars)
+
+        assert ~np.all(
+            self._fit_parameters == 0
+        ), "Either Fits are all zero or assignment of fitting not working"
```
Comment (ajjackson) [inline src/mvesuvio/oop/AnalysisRoutine.py:386]: Is `~np.all(self._fit_parameters == 0)` equivalent to `np.any(self._fit_parameters)`? Maybe I'm missing a subtle distinction.

**Review by ajjackson — COMMENTED** [review]
A few quick comments on Python/numpy stuff, I didn't go through everything!

**Comment by GuiMacielPereira** [comment]
I've changed the vesuvio analysis reduction routine into it's own object, AnalysisReduction. This object is not a Mantid algortihm yet, but I would rather get this PR sorted out first and deal with the transition into a Mantid algorithm separately. Regarding the unit testing, my plan is to follow @MialLewis 's unit tests for the Calibration Mantid algorithm https://github.com/mantidproject/vesuvio/blob/main/tests/unit/calibration/test_calibrate_vesuvio_fit.py. 

@ajjackson What do you think? I suppose turning it into a Mantid algorithm does make the unit tests more verbose, but I don't think that is a good enough reason to avoid doing it?

**Comment by ajjackson** [comment]
I agree that verbosity is not a particularly bad thing in testing. It is good to avoid making tests overly clever/complicated because usually there is no test of the test; sometimes that adds verbosity compared to an "elegant" (with subtle consequences) implementation.

Anyway the patching/mocking in the file you linked to looks like routine stuff, and an appropriate way to implement fine-grained unit tests in code with a lot of interacting methods.

File: src/mvesuvio/util/analysis_helpers.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    mask = (dataX >= start) & (dataX <= end)  # TOF region to mask
+
+    dataY[mask] = 0
+
+    passDataIntoWS(dataX, dataY, dataE, ws)
+    return
+
+
+def extractWS(ws):
+    """Directly exctracts data from workspace into arrays"""
```
Comment (SilkeSchomann) [inline src/mvesuvio/util/analysis_helpers.py:111]:
```suggestion
    """Directly extracts data from workspace into arrays"""
```

File: src/mvesuvio/util/process_inputs.py [hunk 5]
```diff
@@ -81,6 +81,14 @@ def completeICFromInputs(IC, wsIC):
     except AttributeError:
         IC.normVoigt = True
 
+    #Create default for H ratio
+    # Only for completeness sake, will be removed anyway
```
Comment (SilkeSchomann) [inline src/mvesuvio/util/process_inputs.py:85]:
```suggestion
    # Only for completeness' sake, will be removed anyway 
```

File: src/mvesuvio/analysis_reduction.py [hunk 6]
```diff
@@ -1,1114 +1,1053 @@
...
+        self._kinematic_arrays = kinematicArrays
+        self._y_space_arrays = ySpacesForEachMass
 
-    maskBinsWithZeros(wsCrop, ic)  # Used to mask resonance peaks
 
-    MaskDetectors(Workspace=wsCrop, WorkspaceIndexList=ic.maskedDetectorIdx)
-    return wsCrop
+    def run(self):
 
+        assert len(self.profiles) > 0, "Add profiles before atempting to run the routine!"
```
Comment (SilkeSchomann) [inline src/mvesuvio/analysis_reduction.py:238]:
```suggestion
        assert len(self.profiles) > 0, "Add profiles before attempting to run the routine!"
```

File: src/mvesuvio/analysis_reduction.py [hunk 7]
```diff
@@ -1,1114 +1,1053 @@
...
-    histWidths = dataX[:, 1:] - dataX[:, :-1]
-    assert np.min(histWidths) == np.max(
-        histWidths
-    ), "Histogram widhts need to be the same length"
+    def _fit_neutron_compton_profiles(self):
+        """
+        Performs the fit of neutron compton profiles to the workspace being fit.
+        The profiles are fit on a spectrum by spectrum basis.
+        """
+        print("\nFitting Neutron Compron Prolfiles:\n")
```
Comment (SilkeSchomann) [inline src/mvesuvio/analysis_reduction.py:321]:
```suggestion
        print("\nFitting Neutron Compton Prolfiles:\n")
```

File: src/mvesuvio/analysis_reduction.py [hunk 8]
```diff
@@ -1,1114 +1,1053 @@
...
-    dataEf = dataE[~zerosMask]
+        return np.sum((ncpTotal - dataY) ** 2 / dataE**2)
 
-    if np.all(dataE == 0):  # When errors not present
-        return np.sum((ncpTotal - dataYf) ** 2)
 
-    return np.sum((ncpTotal - dataYf) ** 2 / dataEf**2)
+    def _neutron_compton_profiles(self, pars):
+        """
+        Neutron Compron Profile distribution on TOF space for a single spectrum.
```
Comment (SilkeSchomann) [inline src/mvesuvio/analysis_reduction.py:647]:
```suggestion
        Neutron Compton Profile distribution on TOF space for a single spectrum. 
```

File: src/mvesuvio/analysis_reduction.py [hunk 9]
```diff
@@ -1,1114 +1,1053 @@
...
 
-        if np.all(specFitPars == 0):
-            print("Skipped spectra.")
-        else:
-            with np.printoptions(
-                suppress=True, precision=4, linewidth=200, threshold=sys.maxsize
-            ):
-                print(specFitPars)
+    def convertDataXToYSpacesForEachMass(self, dataX, delta_Q, delta_E):
+        "Calculates y spaces from TOF data, each row corresponds to one mass"
```
Comment (SilkeSchomann) [inline src/mvesuvio/analysis_reduction.py:405]:
```suggestion
        """Calculates y spaces from TOF data, each row corresponds to one mass"""
```

File: src/mvesuvio/analysis_reduction.py [hunk 10]
```diff
@@ -1,1114 +1,1053 @@
...
 
-    meanWidths = np.nanmean(betterWidths, axis=1)
-    stdWidths = np.nanstd(betterWidths, axis=1)
+        assert np.all(meanWidths != np.nan), "At least one mean of widths is nan!"
+        assert np.sum(filterMask) >= 1, "No widths survive filtering condition"
+        assert not (np.all(np.isnan(betterWidths))), "All filtered widths are nan"
+        assert not (np.all(np.isnan(betterIntensities))), "All filtered intensities are nan"
+        assert np.nanmax(betterWidths) != np.nanmin(
+            betterWidths
+        ), f"All fitered widths have the same value: {np.nanmin(betterWidths)}"
```
Comment (SilkeSchomann) [inline src/mvesuvio/analysis_reduction.py:573]:
```suggestion
        ), f"All filtered widths have the same value: {np.nanmin(betterWidths)}"
```

File: src/mvesuvio/analysis_reduction.py [hunk 11]
```diff
@@ -1,1114 +1,1053 @@
...
+        assert np.nanmax(betterIntensities) != np.nanmin(
+            betterIntensities
+        ), f"All fitered widths have the same value: {np.nanmin(betterIntensities)}"
```
Comment (SilkeSchomann) [inline src/mvesuvio/analysis_reduction.py:576]:
```suggestion
        ), f"All filtered widths have the same value: {np.nanmin(betterIntensities)}"
```

File: src/mvesuvio/analysis_reduction.py [hunk 12]
```diff
@@ -1,1114 +1,1053 @@
...
+        ncp = self._fit_profiles_workspaces['total'].extractY()
+
+        self._zero_columns_boolean_mask = np.all(dataY == 0, axis=0)  # Masked Cols
+
+        self._workspace_for_corrections = CloneWorkspace(
+            InputWorkspace=self._workspace_for_corrections.name(), 
+            OutputWorkspace=self._workspace_for_corrections.name() + "_CorrectionsInput"
+        )
+        for row in range(self._workspace_for_corrections.getNumberHistograms()):
+            # TODO: Once the option to chage point to hist is removed, remove [:len(ncp)]
```
Comment (SilkeSchomann) [inline src/mvesuvio/analysis_reduction.py:842]:
```suggestion
            # TODO: Once the option to change point to hist is removed, remove [:len(ncp)]
```

File: src/mvesuvio/analysis_reduction.py [hunk 13]
```diff
@@ -1,1114 +1,1053 @@
...
-            meanIntensityRatios /= np.sum(meanIntensityRatios)
+    def calcMSCorrectionSampleProperties(self, meanWidths, meanIntensityRatios):
+        masses = [p.mass for p in self._profiles.values()]
 
-    MSProperties = np.zeros(3 * len(masses))
-    MSProperties[::3] = masses
-    MSProperties[1::3] = meanIntensityRatios
-    MSProperties[2::3] = meanWidths
-    sampleProperties = list(MSProperties)
+        # If Backsscattering mode and H is present in the sample, add H to MS properties
```
Comment (SilkeSchomann) [inline src/mvesuvio/analysis_reduction.py:925]:
```suggestion
        # If Backscattering mode and H is present in the sample, add H to MS properties
```

File: src/mvesuvio/analysis_reduction.py [hunk 14]
```diff
@@ -1,1114 +1,1053 @@
...
+    #
+    #             try:    # Skip floats 
+    #                 float(word) 
+    #             except ValueError: 
+    #                 continue
+    #
+    #             profile, attribute = word
+    #             words[i] = f"pars[{profile_order.index(profile) + attribute_order.index(attribute)}]" 
+    #
+    #     return eval(f"lambda pars: {' '.join(words)}")
```
Comment (SilkeSchomann) [inline src/mvesuvio/analysis_reduction.py:820]: Why not uncomment it now and just not call these functions?

**Review by SilkeSchomann — CHANGES_REQUESTED** [review]

File: src/mvesuvio/analysis_reduction.py (hunk 14 above)
Comment (ajjackson) [reply src/mvesuvio/analysis_reduction.py]: If it's not tested, better to delete it and recover from version control history later!

**Comment by SilkeSchomann** [comment]
@GuiMacielPereira I think with your last commit you have re-introduced all the typos again!

**Review by SilkeSchomann — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**